lib/ex_doc/autolink.ex

```diff
@@ -442,54 +450,79 @@ defmodule ExDoc.Autolink do
     @otpdocs <> "#{module}.html"
   end
 
-  defp local_url(:type, name, arity, config, _original_text) when {name, arity} in @basic_types do
+  defp local_url(kind, name, arity, config, original_text, options \\ [])
+
+  defp local_url(:type, name, arity, config, _original_text, _options)
+       when {name, arity} in @basic_types do
     ex_doc_app_url(Kernel, config) <> "typespecs" <> config.ext <> "#basic-types"
   end
 
-  defp local_url(:type, name, arity, config, _original_text)
+  defp local_url(:type, name, arity, config, _original_text, _options)
        when {name, arity} in @built_in_types do
     ex_doc_app_url(Kernel, config) <> "typespecs" <> config.ext <> "#built-in-types"
   end
 
-  defp local_url(kind, name, arity, config, original_text) do
+  defp local_url(kind, name, arity, config, original_text, options) do
     module = config.current_module
     ref = {kind, module, name, arity}
+    mode = Keyword.get(options, :mode, :regular_link)
+    visibility = Refs.get_visibility(ref)
 
-    cond do
-      Refs.public?(ref) -> fragment(tool(module), kind, name, arity)
-      kind == :function -> try_autoimported_function(name, arity, config, original_text)
-      true -> nil
+    case {kind, visibility} do
+      {_kind, :public} ->
+        fragment(tool(module), kind, name, arity)
+
+      {:function, _visibility} ->
+        try_autoimported_function(name, arity, mode, config, original_text)
+
+      {:type, :hidden} ->
+        nil
+
+      _ ->
+        unless kind == :type and name in [:required, :optional] and arity == 1 do
+          maybe_warn(ref, config, visibility, %{original_text: original_text})
+        end
+
+        nil
```

Comment:
```suggestion
      # skip `@type %{required(...), optional(...), ...}`
      {:type, visibility} when name in [:required, :optional] and arity == 1 ->
        nil

      _ ->
        maybe_warn(ref, config, visibility, %{original_text: original_text})
        nil
```

Reply:
Much better. Done.

Reply:
I don't see this change applied, did you accidentally overwrite your GitHub changes when rebasing and force pushing?

Reply:
Sorry, since I rebased, i had to force push them. They are up now.